File: javascript/src/obs.ts
import { spawn } from "node:child_process";
import { setTimeout as delay } from "node:timers/promises";
import OBSWebSocket from "obs-websocket-js";
import type { ObsConnectionOptions, ObsProcessHandle, ObsProcessOptions, ObsSessionController } from "./types.js";

async function stopObsActionSafely(action: () => Promise<unknown>) {
  try {
    await action();
  } catch {
    // OBS may already be stopped or idle.
  }
}

export async function launchObsProcess(options: ObsProcessOptions = {}): Promise<ObsProcessHandle> {
  const obsCommand = options.obsCommand ?? "obs";
  const obsArgs = options.obsArgs ?? ["--minimize-to-tray", "--disable-shutdown-check"];
  const useXvfb = options.useXvfb ?? true;
  const command = useXvfb ? options.xvfbCommand ?? "xvfb-run" : obsCommand;
  const args = useXvfb
    ? ["-a", "-s", options.xvfbScreen ?? "-screen 0 1920x1080x24", obsCommand, ...obsArgs]
    : obsArgs;

  const childProcess = spawn(command, args, {
    cwd: options.cwd,
    env: options.env ?? globalThis.process.env,
    stdio: "ignore",
  });

  await delay(options.startupDelayMs ?? 4_000);

  return {
    process: childProcess,
    stop: async () => {
      if (childProcess.killed || childProcess.exitCode != null) return;
      childProcess.kill("SIGTERM");
      await delay(1_000);
      if (childProcess.exitCode == null) {
        childProcess.kill("SIGKILL");
      }
    },
  };
}

export async function connectObs(options: ObsConnectionOptions = {}): Promise<ObsSessionController> {
  const client = new OBSWebSocket();
  const address = options.address ?? "ws://127.0.0.1:4455";
  const retryMs = options.connectRetryMs ?? 1_000;
  const timeoutMs = options.connectTimeoutMs ?? 30_000;
  const startedAt = Date.now();

  while (true) {
    try {
      await client.connect(address, options.password);
      break;
    } catch (error) {
      if (Date.now() - startedAt >= timeoutMs) {
        throw error;
      }
      await delay(retryMs);
    }
  }

  return {
    client,
    setScene: async (sceneName: string) => {
      await client.call("SetCurrentProgramScene", { sceneName });
    },
    setBrowserSourceUrl: async (inputName: string, url: string, width = 1920, height = 1080) => {
      await client.call("SetInputSettings", {
        inputName,
        inputSettings: {
          url,
          width,
          height,
        },
        overlay: true,
      });
    },
    startRecord: async () => {
      await client.call("StartRecord");
    },
    stopRecord: async () => {
      await stopObsActionSafely(() => client.call("StopRecord"));
    },
    startStream: async () => {
      await client.call("StartStream");
    },
    stopStream: async () => {
      await stopObsActionSafely(() => client.call("StopStream"));
    },
    disconnect: async () => {
      await client.disconnect();
    },
  };
}

File: python/src/moltchess_content/obs.py
from __future__ import annotations

import subprocess
import time
from dataclasses import dataclass

from obsws_python import ReqClient

from .types import ObsConnectionOptions, ObsProcessOptions


@dataclass(slots=True)
class ObsProcessHandle:
    process: subprocess.Popen[bytes]

    def stop(self) -> None:
        if self.process.poll() is not None:
            return
        self.process.terminate()
        try:
            self.process.wait(timeout=1)
        except subprocess.TimeoutExpired:
            self.process.kill()
            self.process.wait(timeout=1)


@dataclass(slots=True)
class ObsController:
    client: ReqClient

    def set_scene(self, scene_name: str) -> None:
        self.client.set_current_program_scene(scene_name)

    def set_browser_source_url(self, input_name: str, url: str, width: int = 1920, height: int = 1080) -> None:
        self.client.set_input_settings(
            input_name,
            {
                "url": url,
                "width": width,
                "height": height,
            },
            True,
        )

    def start_record(self) -> None:
        self.client.start_record()

    def stop_record(self) -> None:
        try:
            self.client.stop_record()
        except Exception:
            pass

    def start_stream(self) -> None:
        self.client.start_stream()

    def stop_stream(self) -> None:
        try:
            self.client.stop_stream()
        except Exception:
            pass


def launch_obs_process(options: ObsProcessOptions | None = None) -> ObsProcessHandle:
    config = options or ObsProcessOptions()
    if config.use_xvfb:
        command = [
            config.xvfb_command,
            "-a",
            "-s",
            config.xvfb_screen,
            config.obs_command,
            *config.obs_args,
        ]
    else:
        command = [config.obs_command, *config.obs_args]

    process = subprocess.Popen(command, stdout=subprocess.DEVNULL, stderr=subprocess.DEVNULL)
    time.sleep(config.startup_delay_ms / 1000)
    return ObsProcessHandle(process=process)


def connect_obs(options: ObsConnectionOptions | None = None) -> ObsController:
    config = options or ObsConnectionOptions()
    client = ReqClient(
        host=config.address,
        port=config.port,
        password=config.password or "",
    )
    return ObsController(client=client)
